File: examples/python/i2lcd/pots.py
import lcdconst
from pca9535 import PCA9535

class Potentiometer(object):
    '''
    Implementation of Up/Down protocol for MCP401x electronic potentiometers
    used for contrast/backlight control
    '''
    def __init__(self, interface, cs, ud):
	self.interface = interface
	direction = self.interface.getPortDir(lcdconst.CPORT) & ~(cs | ud)
	self.interface.setPortDir(lcdconst.CPORT, direction)
	self.control = self.interface.getPortOutput(lcdconst.CPORT) | cs | ud
	self.interface.setPortOutput(lcdconst.CPORT, self.control)
	self.current = 0x1F
	self.csb = cs
	self.udb = ud

    def cs(self, value):
	self.control = self.control | self.csb if value else self.control & ~self.csb
	self.interface.setPortOutput(lcdconst.CPORT, self.control)

    def ud(self, value):
	self.control = self.control | self.udb if value else self.control & ~self.udb
	self.interface.setPortOutput(lcdconst.CPORT, self.control)

    def power(self, power):
	if power:
	    direction = self.interface.getPortDir(lcdconst.CPORT) & ~(self.csb | self.udb)
	    self.interface.setPortDir(lcdconst.CPORT, direction)
	    self.interface.setPortOutput(lcdconst.CPORT, self.csb)
	    self.set(0x00)

    def inc(self):
	self.current += 1 if self.current < 0x3F else 0
	self.ud(1)
	self.cs(0)
	self.ud(0)
	self.ud(1)
	self.cs(1)


    def dec(self):
	self.current -= 1 if self.current > 0 else 0
	self.ud(0)
	self.cs(0)
	self.ud(1)
	self.ud(0)
	self.cs(1)

    def set(self, value):
	value %= 64
	diff = 0
	if value > self.current:
	    diff = value - self.current
	    self.ud(1)
	    self.cs(0)
	    for i in range(diff):
		self.ud(0)
		self.ud(1)
	    self.cs(1)
	    if (self.current + diff) > 0x3F:
		self.current = 0x3F
	    else:
		self.current += diff
	elif value < self.current:
	    diff = self.current - value
	    self.ud(0)
	    self.cs(0)
	    for i in range(diff):
		self.ud(1)
		self.ud(0)
	    self.cs(1)
	    if (self.current - diff) < 0:
		self.current = 0
	    else:
		self.current = self.current - diff

File: examples/cpp-linux/pots.cpp

#include <unistd.h>
#include <time.h>
#include <pots.h>

using namespace i2lcd;

/**
 * @brief Helper function to wait in ns time spans. In average CPUs it is not
 * really necessary, but on faster CPUs no using such delay, may cause
 * communication problems with the chips.
 */
inline int nsleep(long nanos)
{
    struct timespec ts, rm;
    ts.tv_sec = 0;
    ts.tv_nsec = nanos;
    return nanosleep(&ts, &rm);
}

/**
 * @brief Class constructor
 * @param PCA9535 chip class for communication with pots.
 * @param cs - line number where CS line of MCP401x is connected to
 * @param ud - line number where UD line of MCP401x is connected to
 **/
Potentiometer::Potentiometer(PCA9535 &chip, t_PotBits cs, t_PotBits ud) : current(0), csb((uint8_t) cs), udb((uint8_t) ud), iface(chip)
{
    uint8_t tmp;

    tmp = iface.getDirection(CPORT);
    iface.setDirection(CPORT, tmp & (~(csb | udb)));

    control = iface.getOutput(CPORT);
    iface.setOutput(CPORT, control | csb | udb);

    for(tmp=0; tmp < 64; tmp++)
	dec();


}

/**
 * @brief Class destructor
 **/
Potentiometer::~Potentiometer()
{
    uint8_t tmp;

    set(0x00);
    tmp = iface.getDirection(CPORT);
    iface.setDirection(CPORT, tmp | csb | udb);
}

/**
 * @brief Set state of CS line
 * @param True for 1, False for 0
 **/
void Potentiometer::cs(bool value)
{
    control = value ? (control | csb) : (control & (~csb));
    iface.setOutput(CPORT, control);
}

/**
 * @brief Set state of UD line
 * @param True for 1, False for 0
 **/
void Potentiometer::ud(bool value)
{
    control = value ? (control | udb) : (control & (~udb));
    iface.setOutput(CPORT, control);
}

/**
 * @brief Decrement current value of potentiometer by 1
 **/
void Potentiometer::dec()
{
    current -= current > 0 ? 1 : 0;
    ud(0);
    nsleep(750);
    cs(0);
    ud(1);
    nsleep(500);
    ud(0);
    usleep(5);
    cs(1);
    usleep(5);
}

/**
 * @brief Increment current value of potentiometer by 1
 **/
void Potentiometer::inc()
{
    current += current < 63 ? 1 : 0;
    ud(1);
    nsleep(750);
    cs(0);
    ud(0);
    nsleep(500);
    ud(1);
    usleep(5);
    cs(1);
    usleep(5);
}

/**
 * @brief Set potentiometer to given value.
 *        Will set value of pot by calculating delta between
 *        requested value and current value and depend of sign of the delta
 *        will decrement/increment pot by one until desired value is reached.
 *        These potentiometers don't have any register we can read to get current
 *        wiper position, so we must remeber state of pot in class variable.
 * @param value
 **/
void Potentiometer::set(uint8_t value)
{
    uint8_t diff, i;

    value %= 0x40;

    if (value > current)
    {
	diff = value - current;
	ud(1);
	nsleep(750);
	cs(0);
	for (i=0; i < diff; i++)
	{
	    ud(0);
	    nsleep(500);
	    ud(1);
	    nsleep(500);
	}
	usleep(5);
	cs(1);
	if ((current + diff) > 0x3f)
	    current = 0x3f;
	else
	    current += diff;
    } else if (value < current)
    {
	diff = current - value;
	ud(0);
	nsleep(750);
	cs(0);
	for(i=0; i < diff; i++)
	{
	    ud(1);
	    nsleep(500);
	    ud(0);
	    nsleep(500);
	}
	usleep(5);
	cs(1);
	if ((current - diff) < 0)
	    current = 0;
	else
	    current -= diff;
    }
}
